File: app/src/main/java/com/pin/recommend/adapter/CharactersAdapter.java
package com.pin.recommend.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.pin.recommend.R;
import com.pin.recommend.dialog.DeleteDialogFragment;
import com.pin.recommend.dialog.DialogActionListener;
import com.pin.recommend.model.entity.RecommendCharacter;
import com.pin.recommend.model.viewmodel.CharacterListViewModel;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class CharactersAdapter extends BaseAdapter {

    private AppCompatActivity context;

    private LayoutInflater inflater;

    private List<RecommendCharacter> characters;

    public static final Calendar NOW = Calendar.getInstance();

    private boolean isEditMode = false;

    private CharacterListViewModel characterViewModel;

    public CharactersAdapter(AppCompatActivity context, CharacterListViewModel characterViewModel){
        this.context = context;
        this.inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.characters = new ArrayList<>();
        this.characterViewModel = characterViewModel;
    }

    public void setList(List<RecommendCharacter> list){
        this.characters = list;
        notifyDataSetChanged();
    }

    public void setEditMode(boolean editMode){
        isEditMode = editMode;
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return characters.size();
    }

    @Override
    public RecommendCharacter getItem(int position) {
        return characters.get(position);
    }

    @Override
    public long getItemId(int position) {
        return characters.get(position).id;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View v = convertView;
        //if (v == null){
            v = inflater.inflate(R.layout.row_character, parent,false);
        //}

        final RecommendCharacter character = characters.get(position);

        TextView name = v.findViewById(R.id.character_name);
        name.setText(character.name);

        TextView elapsedTime = v.findViewById(R.id.elapsedTime);
        elapsedTime.setText(character.getDiffDays(NOW));

        TextView created = v.findViewById(R.id.created);
        created.setText(character.getFormattedDate());

        if(character.hasIconImage()){
            ImageView iconImageView = v.findViewById(R.id.character_icon);
            iconImageView.setImageBitmap(character.getIconImage(context, 150, 150));
        }

        ImageView editModeView = v.findViewById(R.id.delete);
        if(isEditMode){
            editModeView.setVisibility(View.VISIBLE);
        }else{
            editModeView.setVisibility(View.GONE);
        }
        editModeView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(isEditMode){
                    DeleteDialogFragment dialog = new DeleteDialogFragment(new DialogActionListener<DeleteDialogFragment>() {
                        @Override
                        public void onDecision(DeleteDialogFragment dialog) {
                            characterViewModel.delete(character);
                        }

                        @Override
                        public void onCancel() {

                        }
                    });
                    dialog.show(context.getSupportFragmentManager(), DeleteDialogFragment.Tag);
                }
            }
        });

        return v;
    }

}

File: app/src/main/java/com/pin/recommend/ui/adapter/CharactersAdapter.kt
package com.pin.recommend.ui.adapter

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.pin.recommend.R
import com.pin.recommend.domain.entity.RecommendCharacter
import com.pin.recommend.ui.character.CharacterListViewModel
import com.pin.recommend.ui.component.DeleteDialogFragment
import com.pin.recommend.ui.component.DialogActionListener
import java.util.Calendar

class CharactersAdapter(
    private val context: AppCompatActivity,
    private val characterViewModel: CharacterListViewModel
) : BaseAdapter() {
    private val inflater: LayoutInflater
    private var characters: List<RecommendCharacter>
    private var isEditMode = false

    init {
        inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        characters = ArrayList()
    }

    fun setList(list: List<RecommendCharacter>) {
        characters = list
        notifyDataSetChanged()
    }

    fun setEditMode(editMode: Boolean) {
        isEditMode = editMode
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return characters.size
    }

    override fun getItem(position: Int): RecommendCharacter {
        return characters[position]
    }

    override fun getItemId(position: Int): Long {
        return characters[position].id
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        val v: View = inflater.inflate(R.layout.row_character, parent, false)
        val character = characters[position]
        val name = v.findViewById<TextView>(R.id.character_name)
        name.text = character.name
        val elapsedTime = v.findViewById<TextView>(R.id.elapsedTime)
        elapsedTime.text = character.getDiffDays(NOW)
        val created = v.findViewById<TextView>(R.id.created)
        created.text = character.formattedDate
        if (character.hasIconImage()) {
            val iconImageView = v.findViewById<ImageView>(R.id.character_icon)
            iconImageView.setImageBitmap(character.getIconImage(context, 150, 150))
        }
        val editModeView = v.findViewById<ImageView>(R.id.delete)
        if (isEditMode) {
            editModeView.visibility = View.VISIBLE
        } else {
            editModeView.visibility = View.GONE
        }
        editModeView.setOnClickListener {
            if (isEditMode) {
                val dialog =
                    DeleteDialogFragment(object :
                        DialogActionListener<DeleteDialogFragment> {
                        override fun onDecision(dialog: DeleteDialogFragment) {
                            characterViewModel.delete(character)
                            val updateWidgetRequest =
                                Intent("android.appwidget.action.APPWIDGET_UPDATE")
                            context.sendBroadcast(updateWidgetRequest)
                        }

                        override fun onCancel() {}
                    })
                dialog.show(context.supportFragmentManager, DeleteDialogFragment.Tag)
            }
        }
        return v
    }

    companion object {
        val NOW = Calendar.getInstance()
    }
}